src/ringmesh/geomodel/geomodel_builder.cpp

```diff
@@ -1006,6 +1032,217 @@ namespace RINGMesh
+            {
+                ringmesh_assert(
+                    line_2_surface[2 * cur_line_id + ( cur_side ? 0 : 1 )]
+                    == NO_ID );
+                line_2_surface[2 * cur_line_id + ( cur_side ? 0 : 1 )] =
+                    surface_count;
+                cur_surface_boundaries.emplace_back( cur_line_id, cur_side );
+
+                const auto& cur_line = this->geomodel_.line( cur_line_id );
+                index_t cur_corner_id{
+                    cur_line.boundary_gmme( cur_side ? 1 : 0 ).index()
+                };
+                const auto& cur_corner =
+                    this->geomodel_.corner( cur_corner_id );
+                double min_angle{ max_float64() };
+                index_t next_line_id{ NO_ID };
+                bool next_side{ false };
+                if( cur_corner.nb_incident_entities() == 1 )
+                {
+                    // Case the current line is an internal border of a surface
+                    next_line_id = cur_line_id;
+                    next_side = !cur_side;
+                }
+                else
+                {
+                    for( auto next_line_itr :
+                        range( cur_corner.nb_incident_entities() ) )
+                    {
+                        const auto& next_line =
+                            cur_corner.incident_entity( next_line_itr );
+                        if( next_line.index() == cur_line_id )
+                        {
+                            continue;
+                        }
+                        auto cur_angle = compute_angle_at_corner(
+                            cur_line, cur_corner, next_line );
+                        if( cur_angle < min_angle )
+                        {
+                            min_angle = cur_angle;
+                            next_line_id = next_line.index();
+                        }
+                    }
+                    ringmesh_assert( next_line_id != NO_ID );
+                    next_side =
+                        this->geomodel_.line( next_line_id ).boundary_gmme( 0 )
+                        == cur_corner.gmme();
+                }
+
+                if( next_line_id == first_line_id && next_side == first_side )
+                {
+                    stop = true;
+                }
+                else
+                {
+                    cur_line_id = next_line_id;
+                    cur_side = next_side;
+                }
+            }
+            surface_boundary_lines.push_back( cur_surface_boundaries );
+            ++surface_count;
+        }
+    }
+
+    void GeoModelBuilder< 2 >::check_internal_floating_lines(
+        const std::vector< index_t >& line_2_surface,
+        const index_t nb_found_surfaces )
+    {
+        std::vector< bool > are_surfaces_hole( nb_found_surfaces, true );
+        for( auto line_id : range( geomodel_.nb_lines() ) )
+        {
+            if( line_2_surface[2 * line_id] != line_2_surface[2 * line_id + 1] )
```

Comment:
I would do the opposite test. If these two are equal, then you have a hole. In your case, I assume that you can put your flag to false and later find out that your surface has an hole. Or maybe I did not really understood the algorithm..